File: java/17.19.java
class Solution {
    public int[] missingTwo(int[] nums) {
        int idx = 0, extend1 = -1, extend2 = -1, n = nums.length;
        while (idx < n) {
            var x = nums[idx];
            if (x == 0 || x == -1) {
                idx++;
                continue;
            }
            
            x = x - 1;
            if (x == n) {
                extend1 = 0;
                nums[idx] = -1;
            } else if (x == n + 1) {
                extend2 = 0;
                nums[idx] = -1;
            } else {
                nums[idx] = nums[x];
                nums[x] = 0;
            }

        }

        int[] ans = new int[2];
        int i = 0;
        for (idx = 0; idx < n; idx++) {
            if (nums[idx] == -1) {
                ans[i] = idx + 1;
                i++;
            }
        }
        if (i < 2 && extend1 == -1) {
            ans[i] = n + 1;
            i++;
        }
        if (i < 2 && extend2 == -1) {
            ans[i] = n + 2;
        }
       
        return ans;
    }
}

File: py/17.19.py
class Solution:
    def missingTwo(self, nums: List[int]) -> List[int]:
        xor = 0
        n = len(nums) + 2
        nums = nums + list(range(1,n+1))
        for i in nums:
            xor ^= i

        bit = xor & ( -xor ) # get the low bit of 1
        l = r = 0
        for i in nums:
            if (bit & i) >= 1:
                l ^= i
            else:
                r ^= i

        return [l, r]

        
        

        